components/SerialServer/src/serial.c

```diff
@@ -498,84 +500,59 @@ void pre_init(void)
     error = camkes_io_ops(&io_ops);
     ZF_LOGF_IF(error, "Failed to initialise IO ops");
 
-    // Initialize the serial port
+    /* Initialize the serial port */
     plat_pre_init(&io_ops);
     set_putchar(serial_putchar);
-    /* query what getchar clients exist */
-    if (getchar_num_badges) {
-        num_getchar_clients = getchar_num_badges();
-        getchar_clients = calloc(num_getchar_clients, sizeof(getchar_client_t));
-        for (int i = 0; i < num_getchar_clients; i++) {
-            unsigned int badge = getchar_enumerate_badge(i);
-            assert(badge <= num_getchar_clients);
-            getchar_clients[i].client_id = badge;
-            getchar_clients[i].buf = getchar_buf(badge);
-            getchar_clients[i].last_head = -1;
-        }
-    }
+
     plat_post_init(&(io_ops.irq_ops));
+
     /* Start regular heartbeat of 500ms */
     timeout_periodic(0, 500000000);
     error = serial_unlock();
 }
 
-void post_init(void)
+int virtqueue_init(void)
 {
-    if (num_registered_virtqueue_channels > 0) {
-        int res = virtqueue_init();
-        if (res) {
-            ZF_LOGE("Serial server does not support read and write virtqueues");
+    int err;
+    /* Query what clients exist */
+    num_clients = ARRAY_SIZE(serial_layout);
+    clients = calloc(num_clients, sizeof(client_t));
+
+    for (int i = 0; i < num_clients; i++) {
+        virtqueue_driver_t *vq_send;
+        virtqueue_device_t *vq_recv;
+
+        vq_recv = malloc(sizeof(*vq_recv));
+        ZF_LOGF_IF(!vq_recv, "Unable to alloc recv camkes-virtqueue for client: %d", i);
```

Comment:
This lacks proper resource handling:
- call free() for everything that has been allocated so far.
- print an error via `ZF_LOGE()` and return an error. It's up to the caller to call `ZF_LOGF_IF()` then.